### ISSUE TITLE
Парсер по автоматической выгрузке данных из data.gov.ru и загрузке в хаб открытых данных

### ISSUE THREAD

**Issue by ivbeg** [infoculture/datatasks#1]
## Цель

Гарантировать сохранение и возможность удобного повторного использования открытых данных публикуемых на портале data.gov.ru
## Задача
- перенести все имеющиеся данные и их описание с портала data.gov.ru на портал hubofdata.ru
- предусмотреть возможность обновления без перегрузки всех данных
## Требования
- открытый исходный код в Github под свободной лицензией
- межплатформенный код (возможность запуска на Linux/Windows)
- предусмотреть перенос всех метаданных в CKAN в виде тегов и атрибутов
- переносить все данные в CKAN или на отдельный хостинг (например, в Github)
- необходимо создавать профили организаций в CKAN аналогично профилям организаций на data.gov.ru
## Пожелания
- реализация в виде программы командной строки или веб-приложения
- возможность запуска на MacOSX
- использовать скриптовые языки такие как Python, Perl, R и другие.
## Оценки трудоёмкости

Задача должна занять не более 1 недели.
## Вспомогательные материалы:
- документация API CKAN - http://docs.ckan.org/en/latest/api/index.html
- документация Data.gov.ru - http://data.gov.ru/pravila-i-rekomendacii

**Comment by ziavra** [comment]
Задача актуальна еще? Если да, то планирую заняться и есть несколько вопросов.
1. Часть метаданных недоступна через API, к примеру "паспорт набора данных"
   http://data.gov.ru/opendata/7710474375-svedovak
   или сведения об организации(API возвращает только название организации и ID)
   http://data.gov.ru/pravitelstvo-volgogradskoy-oblasti
   нужно ли дополнительно парсить эти данные или ограничиться возможностями API?
2. планирую добавлять наборы данных и организации с именами вида datagovru_original_name, проверять существование и обновлять в дальнейшем по этому же ключу.
   устроит ли такой вариант? подозреваю, что он может привести к появлению дублей в дальнейшем, учитывая, например, планируемый автоматический импорт данных с data.mos.ru
   т.е. сейчас дубли тоже появятся, но один раз их можно разобрать руками, а как разрешать подобные конфликты автоматом я пока не знаю
3. на исходном сайте кроме тэгов, есть еще "темы"
   http://data.gov.ru/opendata
   их как переносить: просто добавить в тэги или создать для каждой тему отдельную группу? я думал всё сложить в группу datagovru, к примеру, но можно сделать несколько вида datagovru_security etc
4. каким образом добавлять пакеты и организации, чтобы я мог их удалить/отредактировать в процессе отладки? достаточно прописать себя в автора/майнтейнера при создании или настроена система с доступами?

**Comment by ivbeg** [comment]
@ziavra да, задача остается актуальна
1. Если возможно парсить и несложно, то лучше это делать, если сложно, то ненужно.
2. Я бы предложил использовать коды внутри data_gov_ru такие как "7711000890-smi" они состоят из ИНН организации и ключевого слова уникального названия массива в организации.
   Или сделать короткий префикс "dgr_"
3. Лучше всего темы превращать в тэги.
4. Да, надо создать организации и пакеты от имени пользователя и по его API ключу

**Comment by smartynov** [comment]
Иван, а в чем суть идеи, почему надо брать данные из одного источника и переносить в другое место?

**Comment by ziavra** [comment]
Добрался до задачи в очередной раз, возникла проблема на этапе удаления организаций(и групп, которые автоматически создаются при добавлении организации). Обычное удаление(через web-морду и API) выставляет флаг deleted, но из базы не удаляет. Есть функции organization_purge и group_purge, но у меня не хватает прав для их использования, судя по всему -  получаю ошибку "отказано в доступе"

**Comment by ivbeg** [comment]
Я смогу удалить организации из базы, это делается вручную. Пока предлагаю создать другие, с другими ID'шниками

**Comment by ziavra** [comment]
В процессе выяснилось, что уникальные коды набора данных, которые у них как раз и состоят из ИНН+англоязычного названия, не очень уникальны. От чего ломается, во-первых, моя идея - я хотел именно по этим кодам проверять наличие такого набора и необходимость его обновления, а, во-вторых, ломается их API - ряд наборов просто недоступны, выдаётся только один, либо файлы от другогого. Вот думаю, что с этим делать: создать руками список исключений и игнорировать часть данных(всего таких 19 штук получается), либо обрабатывать каким-то специальным образом, либо что-то еще. Письмо в техподдержку напишу, конечно, но писал им по ошибкам в документации на сайте почти 2 недели назад - ни ответа, ни исправлений.

**Comment by ziavra** [comment]
Возникли некоторые сложности с загрузкой файлов. Я использую для работы [рекомендованное API](https://github.com/ckan/ckanapi) - всё работает отлично, но при загрузке файлов выдаёт ошибку. попробовал пример из [документации](http://ckan.readthedocs.org/en/latest/maintaining/filestore.html) - аналогичная картина с той же ошибкой. [Расследование](https://lists.okfn.org/pipermail/ckan-dev/2014-February/006905.html) показало, что на сайте установлена версия ckan 2.2a, а с тех пор загрузку файлов целиком переписали. Попробовал [ckanclient](https://github.com/okfn/ckanclient-deprecated), который depricated, с ним файл добавился. 
Хотел посоветоваться насчёт вариантов. Поднимать версию CKAN ради этого никто не будет, наверное. Соответственно, переписывать всё на старую версию API с использованием ckanclient'a этого или оставлять как есть, но загрузку файлов делать этим клиентом, или вообще попробовать код из него вытащить и переписать самому?

**Comment by ziavra** [comment]
В итоге сделал я черновой вариант. Дубли пока просто игнорируются, складываются в отдельный файлик при запуске. Работает с новым API, для аплоада файлов использует старый ckanclient. Добавил 10 наборов для теста
http://hubofdata.ru/group/data-gov-ru
Если такой формат устраивает и замечаний нет, то доделаю обновление данных и выложу код в открытый репозитарий на github. Сейчас код в приватном репозитарии на битбакете, могу сделать доступ, если требуется.
Еще вопрос по лицензии, какую лицензию для данных нужно ставить?

**Comment by ivbeg** [comment]
Да, вполне устраивает. Очень даже круто! Лицензию ставим везде "Other", в общем не свободная там лицензия

**Comment by ziavra** [comment]
Дописал, выложил код.
https://github.com/ziavra/datagov2hubofdata

Могу запустить импорт данных. Из-за невозможности полного удаления данных я тестировал добавление только на небольшом количестве организаций/пакетов и хотелось бы первоначальный импорт проводить при моём непосредственном участии, чтобы иметь возможность оперативно исправлять баги.
